### PR TITLE
reduce chunk_size in scan for huge length types

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [x] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Reduce the chunk_size in Scan for huge length types, including JSON, HLL, BITMAP.

### CODE REVIEW

File: be/src/runtime/types.h [hunk 1]
```diff
@@ -261,55 +261,9 @@ struct TypeDescriptor {
...
-        case TYPE_CHAR:
-        case TYPE_STRUCT:
-        case TYPE_TIME:
-            DCHECK(false);
-        }
-        return 0;
-    }
+    // For some types with potential huge length, whose memory consumption is far more than normal types,
+    // they need a different chunk_size setting
+    bool is_huge_type() const { return len >= 128; }
```
Comment (imay) [inline be/src/runtime/types.h:266]: What `len` is for HLL and JSON type.

File: be/src/exec/pipeline/olap_chunk_source.cpp [hunk 2]
```diff
@@ -173,12 +188,7 @@ Status OlapChunkSource::_init_reader_params(const std::vector<OlapScanRange*>& k
     _params.profile = _runtime_profile;
     _params.runtime_state = _runtime_state;
     _params.use_page_cache = !config::disable_storage_page_cache;
-    // Improve for select * from table limit x, x is small
-    if (_limit != -1 && _limit < _runtime_state->chunk_size()) {
-        _params.chunk_size = _limit;
-    } else {
-        _params.chunk_size = _runtime_state->chunk_size();
-    }
+    _decide_chunk_size();
```
Comment (imay) [inline be/src/exec/pipeline/olap_chunk_source.cpp:191]: Why not do this in FE when planning?

File: be/src/runtime/types.h (hunk 1 above)
Comment (mofeiatwork) [reply be/src/runtime/types.h:266]: HLL: 16385, JSON: 128.

File: be/src/exec/pipeline/olap_chunk_source.cpp (hunk 2 above)
Comment (mofeiatwork) [reply be/src/exec/pipeline/olap_chunk_source.cpp:191]: `chunk_size` is a global parameter, which affects all operators.
Current bad case is located at scan operator, the IO thread allocates chunk and transfer to computation thread, which cause memory allocation bottleneck.

Current solution is just a work-around, we need a better memory allocator to solve there issues.

Comment (imay) [reply be/src/exec/pipeline/olap_chunk_source.cpp:191]: Oh I see.
But I have another question, if some other ExecNodes will generate `huge` type, will it have the memory allocation problem and affect the performance?
IMO, we can give a smaller(like 4096) global chunk_size as long as there is a `huge` type in the whole query.

Comment (mofeiatwork) [reply be/src/exec/pipeline/olap_chunk_source.cpp:191]: In theory it has.
And the whole solution may be, each operator needs to decide the chunk size individually, according to its input type and output type.

For this solution, we need to benchmark carefully, which is not easy work. 

And currently we need to focus at the performance degradation between pipeline & non-pipeline mode, and solve this issue at `version-2.2`, and leave the adaptive chunk_size & query-intrinsic memory allocator to future work.

**Review by ZiheLiu — DISMISSED** [review]

File: be/src/runtime/types.h (hunk 1 above)
Comment (kangkaisen) [inline be/src/runtime/types.h:266]: what the bitmap type len?

**Review by silverbullet233 — DISMISSED** [review]

**Comment by mofeiatwork** [comment]
@Mergifyio backport branch-2.2

File: be/src/runtime/types.h (hunk 1 above)
Comment (mofeiatwork) [reply be/src/runtime/types.h:275]: No length previous, so I change it 128, to consider it as a huge type.

**Review by kangkaisen — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

File: be/src/runtime/types.h (hunk 1 above)
Comment (trueeyu) [reply be/src/runtime/types.h:275]: does the varchar type or array type have 'len' ?

Comment (trueeyu) [reply be/src/runtime/types.h:275]:
```
    static TypeDescriptor create_varchar_type(int len) {
        DCHECK_GE(len, 1);
        DCHECK_LE(len, MAX_VARCHAR_LENGTH);
        TypeDescriptor ret;
        ret.type = TYPE_VARCHAR;
        ret.len = len;
        return ret;
    }
```

**Review by imay — APPROVED** [review]

**Pull request merged**